core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -741,9 +741,10 @@ abstract class RDD[T: ClassTag](
   def mapWith[A, U: ClassTag]
       (constructA: Int => A, preservesPartitioning: Boolean = false)
       (f: (T, A) => U): RDD[U] = withScope {
+    val cleanF = sc.clean(f)
     mapPartitionsWithIndex((index, iter) => {
       val a = constructA(index)
-      iter.map(t => f(t, a))
+      iter.map(t => cleanF(t, a))
```

Comment:
@rxin @ted-yu actually even though `mapPartitionsWithIndex` does clean already, it cleans the whole closure but not the ones used in the closure. In this case, I believe it's actually necessary to clean `f` here since we won't actually clean it from `mapPartitionsWithIndex`. For the same reason I believe we also need to clean `constructA` since it's a closure provided by the user.

Please add the corresponding test cases to `ClosureCleanerSuite`.